### ISSUE TITLE
Accessibility: Default focus should not always be on input textarea

### ISSUE THREAD

**Issue by cloudscape-germany** [PrivateBin/PrivateBin#1794]
We've encountered this issue in this community here before: https://github.com/PrivateBin/PrivateBin/issues/1386

The issue is, that the immeadiate focus on one specific part of the web page might save healthy people a few seconds but will disoriented reading/vision imparied people to a great degree. This would make PrivateBin fail accessibility tests of governments of countries in the EU. This would be a shame, because the entire EU is switching to free and open software away from propietery software and this would be a major show stopper, since governments -have to- adhere to accessibility legislation. The consequences would be bad for all of us Id say.

Thus I want to propose if we could either change this auto-focus feature and make the project more accessible in general in the long run or at least give us the option to rewrite functionality outside of `privatebin.js` file.

**Comment by cloudscape-germany** [comment]
Did you actually read my message?

**Comment by rugk** [comment]
Indeed, I see the difference here: https://github.com/PrivateBin/PrivateBin/issues/1386 was about the focus being _trapped_ in the text area. This issue, however, is about the **default focus**.

To elaborate (@cloudscape-germany you were good explaining the impact, but maybe next time try to explain the actual issue more clearly, IMHO. Also fill out the issue template. Accessibility issues can be filled as bugs.): If you visit the page, your focus is instantly in the input box for a new paste. This is great UX for most users, but I agree this could be an a10y issue for users of screen readers, because AFAIK screen readers always only read what is focused (first)? So they miss important context?

I guess that is the suspected issue.

/cc ing the creator of the other issue @bugnumber9 as you seem to have some a10y knowledge or at least awareness here too.

**Comment by rugk** [comment]
What I know and what you can e.g. see on GitHub too is a solution for the "inverted" problem though: If you visit a new page and press tab you get a "Skip to content box". That is a common a10y pattern, cf.: https://www.w3schools.com/accessibility/accessibility_skip_links.php

So I tried finding information about what default focus is suggested, but unfortunately [most information centers around the appearance of the focus](https://www.w3.org/WAI/WCAG22/Understanding/focus-appearance).
Famous WCAG guidelines e.g. describe things related to navigation here, e.g.: https://www.w3.org/WAI/WCAG22/Understanding/navigable.

Except of:
> provide a way for the user to orient herself within a set of web pages

and [the already explained "bypass blocks"](https://www.w3.org/WAI/WCAG22/Understanding/bypass-blocks) I could find nothing related.

> accessibility legislation

I know in Germany the technical details essentially boil down to "follow the WCAG guidelines". Anyway, obviously, we do not only want a10y for legal reasons, but in general it's of course a good aim to achieve.

> This would make PrivateBin fail accessibility tests of governments of countries in the EU

Is that default focus a reason for that? [Citation needed](https://en.wikipedia.org/wiki/Wikipedia:Citation_needed) I guess. Respectively, as said I tried finding suggestions/guidelines on the default focus, but I could not find anything.
Contrary [the bypass blocks guideline](https://www.w3.org/WAI/WCAG22/Understanding/bypass-blocks) seems to suggest "users should find the main content first". So if some user comes to visit PrivateBin, they likely already know what service it is and want to create a paste, would not they? So focusing on. This is essentially a "bypass box", but defaults to bypassing (and you would need to press (<kbd>shift</kbd>+)<kbd>tab</kbd> to navigate around there).
I actually doubt that on the main page people would want to have all buttons (New etc.) and options to be read/navigate. 

I agree a10y is important, but I guess I do not yet see the concrete problem.

**Comment by elrido** [comment]
> Did you actually read my message?

I did read this and as you yourself admitted, this has already been discussed at length in the original issue and we addressed the other concerns but already concluded that moving the focus to the main input area is in the interest of accessibility and not against it.

I do not see any new evidence presented here that would conclude that helping the users start creating a document by moving the focus to the text input is a problem for accessibility. I'm myself visually impaired and therefore do very much appreciate accessibility and this feature in particular.

**Comment by cloudscape-germany** [comment]
@rugk Our Privatebin fork failed the BITV 2.0 Review due to the initial focus issue described in this issue/bug, please see Sources [1] and [2]. There are about 1 Mio. affected users in total, from which of course not all are visually imparied, but if tests can't be passed, it would affect a multitute of decisions we can't estimate, but it would affect choice of software, employability, thus un/employment numbers as well as give competitors from the propietery software market an advantage. So this is not only a business issue, but also an issue of digital sovereignty. (btw, I think you referred to a11y?)

Id suggest everybody interested in this issue runs an automated BITV 2.0 herself to see if she can see the need to develop a longer term compatibiltiy to the accessibility regulations. I mean we can code our way around it, but in the long run, we all would benefit from the more strategic decisions we make in this regard. 

@elrido Thanks for that feedback, we never tried the screen reader ourselves admittely, but the automated test did test the elements mentioned. I would kindly suggest giving the cretors of the BITV 2.0 testing paradigm feedback on this particular issue to make them reconsider. In our quick tests, more specifically, Ctrl-m muted the audio of our browser and ESC did nothing to the focus behaviour.

1. https://www.gesetze-im-internet.de/bitv_2_0/BJNR184300011.html
2. https://www.bmas.de/DE/Service/Gesetze-und-Gesetzesvorhaben/barrierefreie-informationstechnik-verordnung-2-0.html

**Comment by rugk** [comment]
[1] and [2] are basically the law which is not conrete enough to actually explain the issue. (more on that below)

If your software/fork failed some test, we would be very glad to:
1. get the (full, if possible) report (of the "BITV 2.0 Review") on that. This should provide and give enough insides.
2. and at best, just because I am curious: Is your source code/fork published somewhere? Maybe on https://gitlab.opencode.de/ / https://opencode.de/de given you seem to be creating software for [edited] institutions. Or what's your fork called?

Note you can upload PDFs on GitHub, so in the full open-source spirit I would prefer to get the report out here.

---

Side note (for people curious on the topic, but slightly off-topic), if one wants to learn more about the German law [here is a nice podcast about it](https://rechtsbelehrung.com/134-bfsg/) (in German). What's relevant for the issue at hand: The law is not concrete enough and (kinda obviously) does not explain detailed stuff like "you must have alt text here", but it essentially boils down to "industry standards" (§ 3 BITV explains it as "harmonisierten Normen" = "harmonized norms", but yet again don't get started on legalese texts). 
Which, essentially, is a thing like [WCAG](https://www.w3.org/WAI/WCAG22/Understanding/about). That's why I previously asked about a reference in the _WCAG guidelines_ even though we talk about a German law here. Kinda complicated, I know.

Edited upon request by OP. Sorry.

**Comment by elrido** [comment]
I'd accept an alternative template or an off-by-default configuration option that has this very important feature disabled. As someone living in a neighboring country it does sound very stereotypical to argue that a FOSS software needs to remove an important feature helping visually impaired people, because of some odd German regulation mandates it.

This is not actually about getting the spirit of that codex and helping those that struggle with our tech, which I do fully support, but just blindly follow a rule to get some boxes ticked, because its "the law" (insert Sylvester Stallone as Judge Dredd Meme). I would instead recommend you fight back against that regulation and get it fixed. This is clearly a false positive that this tool reported to you. Get them to fix that tool, please.

**Comment by rugk** [comment]
Also a detail:
> Id suggest everybody interested in this issue runs an automated BITV 2.0 herself

*Where* or *how* could one run such a report? "automated" sounds like a tool, where can we get/use/scan that?

You stay too vague here, we need the detailed error/result of the output of the tool to make sense of it.

**Comment by cloudscape-germany** [comment]
Test has been conducted automatically, the tool is not publically available, the relevant part has been: 

> After Opening of the page, the focus is set directly into the input field. Visually impaired persons will wrongly recognise this as the beginning of the page and miss the objects that come above the input field.

Furthermore it has been complained that upon opening of dialog boxes, the focus is not set directly upon the boxes but one would need to press tab once. After closing of dialogs the focus should be on the invoking button, but I think this can be done on a template-basis.

**Comment by rugk** [comment]
Opened https://github.com/PrivateBin/PrivateBin/issues/1794#issuecomment-3968486674 for the other issue.

> but I think this can be done on a template-basis.

While it possibly can, we'd generally prefer to have all improvments upstream.

Also, that's why I ask to get the full report, so al improvements can be contributed upstream. It makes no sense to only improve a10y in some forks only.

> the tool is not publically available

Well… anyway sounds like a useful tool? Did you thought about releasing it as open-source (respectively urging people to do that) and thus contribute that back to the open-source community?

**Comment by cloudscape-germany** [comment]
In case a certification is needed, only certified BIT testers may test a web project upon its validity to certify validity with EN 301 549.

On bitvtest.de one could request a test, mentioning the open source nature of PrivateBin and demand a waiver of any fees.

However, if no official certification is needed, it is possible for anyone to conduct a manual test with help of this repository: https://github.com/BIK-BITV/BIK-Web-Test and its over 90 manual testing steps.

**Comment by cloudscape-germany** [comment]
Privatebin would have to pass these criteria **where applicable**:

# Web Accessibility Checklist
Source: [lbit.hessen.de](https://lbit.hessen.de/service-beratung/webseiten)

## Images & Graphics

- Alt text for informative images (1–2 sentences, objective)
- Alt text for interactive elements → name the function (e.g. "search")
- Linked images → describe content + name link destination
- Decorative graphics → `alt=""` (empty)
- CAPTCHA → audio version or non-image-based alternative

## Videos

### General
- Videos without audio → media alternative (text or audio description)
- Videos with/without audio → audio description available
- Videos with audio → subtitles available
- Foreign-language videos → translated audio track available
- No flashing or animated elements

### Subtitles
- Synchronized with audio
- Multiple speakers → unique color per person
- Subtitles must not overlap video content
- Display duration long enough to read
- Max. 37 characters per line (recommended)

### Video Player
- Fully keyboard-operable
- Audio description toggleable directly in player
- Subtitles can be shown/hidden
- Audio description & subtitles preserved during conversion
- Playback speed adjustable (recommended)

## Headings

- Use HTML hierarchy (max. 6 levels)
- Maintain correct order (H1 → H2 → H3 …)
- Short and clear

## Tables

- No layout tables (screen reader issue)
- Mobile-optimized
- Correct HTML tags: `<table>`, `<th>`, `<tr>`, `<td>`
- Simple tables → one header per column and row, no merged cells
- Complex tables → `header` and `id` attributes for unambiguous assignment

## Page Titles

- Two-part structure: individual title + general title (e.g. "Page XY – Website Name")
- General part always consistent

## Links

- Descriptive labels (no bare "click here")
- Not highlighted by color alone (+ symbol, bold, or underline)
- Name the link destination (new page / document → mandatory)
- Alt text for icon links → name the function

## Colors & Contrast

- Text/background → min. 7:1
- Text from 18 pt / bold text from 14 pt → min. 4.5:1
- UI components & graphics → min. 3:1
- Text in images & transparent text → check manually (Color Contrast Analyzer)
- Text logos → no minimum contrast required
- Links, required fields, error messages, menu items, headings → not distinguished by color alone (alternatively 3:1 contrast)
- Table row/cell highlighting → bold, border, or 3:1 contrast

## HTML (Accessible Web Pages)

- Foreign-language words marked with `lang` attribute
- Lists coded as `<ul>` / `<ol>` / `<dl>`
- Quotes coded as `<blockquote>` / `<cite>`
- Form fields labeled with `<label>` (placeholder text alone is insufficient)
- Multi-part forms grouped with `<fieldset>`
- Correct heading order (H1 → H6, no levels skipped)
- Warn users about data loss on inactivity (exception: data retained for 20 h)
- Navigation consistent across all pages
- Repeated elements skippable via skip links
- Multiple ways to access content (navigation + search and/or table of contents)
- Fully keyboard-operable (Tab key)
- Focus state visible
- No timing requirements for keyboard use
- Keyboard focus never trapped
- Abbreviations defined with `<abbr title="...">`

**Comment by rugk** [comment]
Sure, okay thanks for the links. To stay on-topic, though I tried searching the linked repo for focus information and these seem to be the only information and they kinda look not relevant for the problem here:
<img width="1203" height="238" alt="Image" src="https://github.com/user-attachments/assets/9c3d92b8-e263-4a92-831f-d01ff9852f44" />

Search link: https://github.com/search?q=repo%3ABIK-BITV%2FBIK-Web-Test%20fokus&type=code

**Comment by cloudscape-germany** [comment]
Argument was that an imparied person would be confused why screenreader does not start reading at top left element of DOM as usual. Recent BITV Tests did not find this subjective as a problem on our side. 

From our point of view we can close this issue and start another one when we are ready to make the entire application more accessible.
